File: app/src/main/java/com/trevorwiebe/trackacow/presentation/activities/EditLotActivity.java
package com.trevorwiebe.trackacow.presentation.activities;

import android.app.DatePickerDialog;
import android.content.Intent;

import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.domain.dataLoaders.cache.holdingLot.InsertHoldingLot;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.lot.QueryLotByLotId;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.lot.UpdateLot;
import com.trevorwiebe.trackacow.data.local.entities.LotEntity;
import com.trevorwiebe.trackacow.data.local.holdingUpdateEntities.HoldingLotEntity;
import com.trevorwiebe.trackacow.domain.utils.Constants;
import com.trevorwiebe.trackacow.domain.utils.Utility;

import java.util.Calendar;

public class EditLotActivity extends AppCompatActivity implements
        QueryLotByLotId.OnLotByLotIdLoaded {

    private LotEntity mSelectedLot;
    private Calendar mCalendar = Calendar.getInstance();
    private DatePickerDialog.OnDateSetListener mDatePicker;

    private TextInputEditText mLotDescription;
    private TextInputEditText mCustomerName;
    private TextInputEditText mDate;
    private TextInputEditText mNotes;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_lot);

        this.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);
        this.getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        String lotId = getIntent().getStringExtra("lotId");

        new QueryLotByLotId(lotId, this).execute(this);

        mLotDescription = findViewById(R.id.edit_lot_description);
        mCustomerName = findViewById(R.id.edit_customer_name);
        mDate = findViewById(R.id.edit_lot_date);
        mNotes = findViewById(R.id.edit_notes);
        Button updateBtn = findViewById(R.id.update_pen_btn);
        Button cancel = findViewById(R.id.cancel_update_pen_btn);
        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        mDate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new DatePickerDialog(EditLotActivity.this,
                        mDatePicker,
                        mCalendar.get(Calendar.YEAR),
                        mCalendar.get(Calendar.MONTH),
                        mCalendar.get(Calendar.DAY_OF_MONTH))
                        .show();
            }
        });

        mDatePicker = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                mCalendar.set(Calendar.YEAR, year);
                mCalendar.set(Calendar.MONTH, month);
                mCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                mDate.setText(Utility.convertMillisToDate(mCalendar.getTimeInMillis()));
            }
        };

        updateBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mCustomerName.length() == 0) {
                    Snackbar.make(v, "Please fill the blanks", Snackbar.LENGTH_LONG).show();
                }else{

                    String lotDescription = mLotDescription.getText().toString();
                    String customerName = mCustomerName.getText().toString();
                    long date = mCalendar.getTimeInMillis();
                    String notes = mNotes.getText().toString();

                    mSelectedLot.setLotName(lotDescription);
                    mSelectedLot.setCustomerName(customerName);
                    mSelectedLot.setDate(date);
                    mSelectedLot.setNotes(notes);

                    if (Utility.haveNetworkConnection(EditLotActivity.this)) {
                        DatabaseReference baseRef = FirebaseDatabase.getInstance().getReference("users").child(FirebaseAuth.getInstance().getCurrentUser().getUid());
                        baseRef.child(LotEntity.LOT).child(mSelectedLot.getLotId()).setValue(mSelectedLot);
                    }else{

                        Utility.setNewDataToUpload(EditLotActivity.this, true);

                        HoldingLotEntity holdingLotEntity = new HoldingLotEntity(mSelectedLot, Constants.INSERT_UPDATE);

                        new InsertHoldingLot(holdingLotEntity).execute(EditLotActivity.this);
                    }

                    new UpdateLot(mSelectedLot).execute(EditLotActivity.this);

                    Intent intent = new Intent();
                    setResult(RESULT_OK, intent);
                    finish();
                }
            }
        });
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return false;
    }

    @Override
    public void onLotByLotIdLoaded(LotEntity lotEntity) {
        mSelectedLot = lotEntity;

        long longDate = mSelectedLot.getDate();
        mCalendar.setTimeInMillis(longDate);

        String date = Utility.convertMillisToDate(longDate);

        mLotDescription.setText(mSelectedLot.getLotName());
        mCustomerName.setText(mSelectedLot.getCustomerName());
        mDate.setText(date);
        mNotes.setText(mSelectedLot.getNotes());

    }

}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/edit_lot/EditLotActivity.kt
package com.trevorwiebe.trackacow.presentation.edit_lot

import androidx.appcompat.app.AppCompatActivity
import android.app.DatePickerDialog.OnDateSetListener
import com.google.android.material.textfield.TextInputEditText
import android.os.Bundle
import com.trevorwiebe.trackacow.R
import android.app.DatePickerDialog
import com.google.android.material.snackbar.Snackbar
import android.content.Intent
import android.os.Build.VERSION
import android.widget.Button
import androidx.activity.viewModels
import com.trevorwiebe.trackacow.domain.models.lot.LotModel
import com.trevorwiebe.trackacow.domain.utils.Utility
import dagger.hilt.android.AndroidEntryPoint
import java.util.*

@AndroidEntryPoint
class EditLotActivity : AppCompatActivity() {

    private val mCalendar = Calendar.getInstance()
    private lateinit var mDatePicker: OnDateSetListener

    private val editLotViewModel: EditLotViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_lot)

        this.supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)
        this.supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        @Suppress("DEPRECATION")
        val lotModel: LotModel? = if(VERSION.SDK_INT <= 33 ){
            intent.getParcelableExtra("lotModel", LotModel::class.java)
        }else{
            intent.getParcelableExtra("lotModel")
        }

        val lotDescriptionTV = findViewById<TextInputEditText>(R.id.edit_lot_description)
        val customerNameTV = findViewById<TextInputEditText>(R.id.edit_customer_name)
        val dateTV = findViewById<TextInputEditText>(R.id.edit_lot_date)
        val notesTv = findViewById<TextInputEditText>(R.id.edit_notes)
        val updateBtn = findViewById<Button>(R.id.update_pen_btn)

        val cancel = findViewById<Button>(R.id.cancel_update_pen_btn)
        cancel.setOnClickListener { finish() }
        dateTV.setOnClickListener {
            DatePickerDialog(
                this@EditLotActivity,
                mDatePicker,
                mCalendar[Calendar.YEAR],
                mCalendar[Calendar.MONTH],
                mCalendar[Calendar.DAY_OF_MONTH]
            )
                .show()
        }

        mDatePicker = OnDateSetListener { view, year, month, dayOfMonth ->
            mCalendar[Calendar.YEAR] = year
            mCalendar[Calendar.MONTH] = month
            mCalendar[Calendar.DAY_OF_MONTH] = dayOfMonth
            dateTV.setText(Utility.convertMillisToDate(mCalendar.timeInMillis))
        }

        updateBtn.setOnClickListener { v ->
            if (customerNameTV.length() == 0 || lotModel == null || lotDescriptionTV.length() == 0) {
                Snackbar.make(v, "Please fill the blanks", Snackbar.LENGTH_LONG).show()
            } else {
                lotModel.lotName = lotDescriptionTV.text.toString()
                lotModel.customerName = customerNameTV.text.toString()
                lotModel.date = mCalendar.timeInMillis
                lotModel.notes = notesTv.text.toString()

                editLotViewModel.onEvent(EditLotUiEvent.OnLotUpdated(lotModel))

                val intent = Intent()
                setResult(RESULT_OK, intent)
                finish()
            }
        }

        if(lotModel != null) {
            val longDate = lotModel.date
            mCalendar.timeInMillis = longDate
            val date = Utility.convertMillisToDate(longDate)
            lotDescriptionTV.setText(lotModel.lotName)
            customerNameTV.setText(lotModel.customerName)
            dateTV.setText(date)
            notesTv.setText(lotModel.notes)
        }

    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return false
    }
}